### PR TITLE
2048 game

### PR DESCRIPTION
**Pull request by Mi1ind**
Modifying the bevy_2048 game to work with rune.

Key Commands are located in [moving_direction.rs](https://github.com/hotg-ai/rune/blob/747451b60c9fb6d9d7a2c79e620247a8d0894d38/examples/bevy_2048/src/movement/moving_direction.rs#L86-L92) and need to be modified to use rune output.

### CODE REVIEW

**Comment by Michael-F-Bryan** [comment]
I'm guessing we'll need to swap out [this](https://github.com/hotg-ai/rune/blob/747451b60c9fb6d9d7a2c79e620247a8d0894d38/examples/bevy_2048/src/movement/moving_input.rs#L20-L33) with something which grabs the last 1 second of audio from your laptop's microphone. That'd mean we skip the `TryFrom<KeyCode>` impl for `MovingDirection` altogether.

I'm not 100% sure how you read samples from an audio device. Bevy probably already something like this, otherwise you may need to spin up a thread at startup, make it read audio data in the background, and use an `Arc<Mutex<Vec<i16>>>` so the audio thread can pass samples to the main event loop.

**Comment by Michael-F-Bryan** [comment]
I did a bit of digging and [the `cpal` crate](https://docs.rs/cpal/) is probably your best bet for working with audio input devices. It looks like they are set up to manage the threading for you, so we'll just need to make sure we can pass samples back to the game's event loop.

**Comment by Michael-F-Bryan** [comment]
We are currently running into issues where `cpal` asks the OS for 1-channel/16-bit/16 kHz audio but the mac says it only supports 2-channel/16-bit/44.1 kHz audio.

**Comment by Mi1ind** [comment]
@Michael-F-Bryan, I was able to get the audio input config working. macOS seems to allow single channel inputs, but the sample rate is restricted. Also the datatype used to store audio in macOS is f32 (an i16 buffer doesn't work at all).

**Comment by Michael-F-Bryan** [comment]
@Mi1ind, I've just rebased this branch onto `master`. Now that microspeech is working a lot better, should we do whatever we need to make CI pass then merge it in? That way we'll have an example of embedding Rune into a real world application.

@kthakore, where would this sit in our list of priorities?

**Comment by Mi1ind** [comment]
@Michael-F-Bryan, I think we ca go ahead and merge with master once we get the CI to pass.

File: examples/bevy_2048/src/movement/mod.rs [hunk 1]
```diff
@@ -205,10 +207,19 @@ impl Output for Serial {
...
+//     string: Cow<'a, str>,
+// }
+// {"type_name":"&str","channel":2,"elements":["up"],"dimensions":[1]}
+#[derive(Debug, serde::Deserialize)]
 struct Message<'a> {
     type_name: Cow<'a, str>,
     channel: usize,
-    string: Cow<'a, str>,
-}
+    elements: Cow<'a, [str; 1]>,
```
Comment (Mi1ind) [inline examples/bevy_2048/src/movement/mod.rs:222]: @Michael-F-Bryan I'm unable to figure out what to change for the game to read the element output from the rune.

**Comment by Michael-F-Bryan** [comment]
I'm going to close this PR as stale.

**Pull request closed without merging**